apps/bestofjs-nextjs/src/server/api.tsx

```diff
@@ -0,0 +1,81 @@
+import { createHallOfFameAPI } from "./api-hall-of-fame";
+import { createProjectsAPI } from "./api-projects";
+import { createTagsAPI } from "./api-tags";
+import {
+  Data,
+  FETCH_ALL_PROJECTS_URL,
+  getFeaturedRandomList,
+  getProjectId,
+  getTagsByKey,
+  populateProject,
+} from "./api-utils";
+
+type RawData = {
+  projects: BestOfJS.RawProject[];
+  tags: BestOfJS.RawTag[];
+  date: string;
+};
+
+export function createAPI() {
+  async function getData() {
+    return await fetchData();
+  }
+
+  async function fetchData() {
+    const { projects, tags: rawTags, date } = await fetchProjectData();
+    const tagsByKey = getTagsByKey(rawTags, projects);
+    const projectsBySlug: Data["projectsBySlug"] = {};
+    projects.forEach((project) => {
+      projectsBySlug[getProjectId(project)] = project;
+    });
+    const featuredProjectIds = getFeaturedRandomList(projects);
+
+    return {
+      projectCollection: projects,
+      featuredProjectIds,
+      tagCollection: Object.values(tagsByKey),
+      populate: populateProject(tagsByKey),
+      tagsByKey,
+      projectsBySlug,
+      lastUpdateDate: new Date(date),
+    };
+  }
+
+  const context = {
+    getData,
+  };
+
+  const projectsAPI = createProjectsAPI(context);
+  const tagsAPI = createTagsAPI(context);
+  const hallOfFameAPI = createHallOfFameAPI(context);
+
+  return {
+    projects: projectsAPI,
+    tags: tagsAPI,
+    hallOfFame: hallOfFameAPI,
+  };
+}
+
+export const api = createAPI();
+
+async function fetchProjectData(): Promise<RawData> {
+  try {
+    const data = await fetchDataFromRemoteJSON();
+    return data as RawData;
+  } catch (error) {
+    console.error("Unable to fetch data!", (error as Error).message);
+    throw error;
+  }
+}
+
+function fetchDataFromRemoteJSON() {
+  const url = FETCH_ALL_PROJECTS_URL + `/projects.json`;
```

Comment:
This is where we get all of the alerts from when building right?

Can't we cache this call after running it once?

Or better yet do it in `fetchProjectData()`

Reply:
Caching is tricky as I realized the previous solutions caused issues in the real site.

The best would be to be able to reproduce in local what Vercel does in the live site with `fetch` calls 🤔